src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -341,27 +348,92 @@ impl Visitor<'tcx> for Validator<'_, 'mir, 'tcx> {
-                self.visit_projection(&place.base, proj, ctx, location);
-            } else {
-                self.super_rvalue(rvalue, location);
+                self.visit_place_base(&borrowed_place.base, ctx, location);
+                self.visit_projection(&borrowed_place.base, inner, ctx, location);
+                return;
             }
-        } else {
-            self.super_rvalue(rvalue, location);
-        }
 
-        match *rvalue {
-            Rvalue::Use(_) |
-            Rvalue::Repeat(..) |
-            Rvalue::UnaryOp(UnOp::Neg, _) |
-            Rvalue::UnaryOp(UnOp::Not, _) |
-            Rvalue::NullaryOp(NullOp::SizeOf, _) |
-            Rvalue::CheckedBinaryOp(..) |
-            Rvalue::Cast(CastKind::Pointer(_), ..) |
-            Rvalue::Discriminant(..) |
-            Rvalue::Len(_) |
-            Rvalue::Ref(..) |
-            Rvalue::Aggregate(..) => {}
+            // It is *always* legal to borrow a `static` from another `static`, even if it contains
+            // interior mutability. Accessing a `static` from a `const` is disallowed by other
+            // checks.
+            | Rvalue::Ref(_, BorrowKind::Shared, ref borrowed_place)
+            | Rvalue::Ref(_, BorrowKind::Shallow, ref borrowed_place)
+            | Rvalue::Ref(_, BorrowKind::Unique, ref borrowed_place)
+            if matches!(borrowed_place.base, PlaceBase::Static(_))
+            => {}
+
+            // Check for shared borrows of values with interior mutability.
+            | Rvalue::Ref(_, kind @ BorrowKind::Shared, ref borrowed_place)
+            | Rvalue::Ref(_, kind @ BorrowKind::Shallow, ref borrowed_place)
+            | Rvalue::Ref(_, kind @ BorrowKind::Unique, ref borrowed_place)
+            => {
+                // FIXME: Change the `in_*` methods to take a `FnMut` so we don't have to manually
+                // seek the cursors beforehand.
+                self.qualifs.has_mut_interior.cursor.seek_before(location);
+                self.qualifs.indirectly_mutable.seek(location);
+
+                let borrowed_place_has_mut_interior = HasMutInterior::in_place(
+                    &self.item,
+                    &|local| self.qualifs.has_mut_interior_eager_seek(local),
+                    borrowed_place.as_ref(),
+                );
+
+                if borrowed_place_has_mut_interior {
+                    let src_derived_from_illegal_borrow = borrowed_place
+                        .as_local()
+                        .map_or(false, |local| self.derived_from_illegal_borrow.contains(local));
+
+                    // Don't emit errors for borrows of values that are *themselves* the result of
+                    // an illegal borrow (e.g., the outermost `&` in `&&Cell::new(42)`). We want to
+                    // point the user to the place where the original illegal borrow occurred, not
+                    // to subsequent borrows of the resulting value.
```

Comment:
I thought this was resolved by simply not having any qualifs on `&Cell::new(42)`.

`&T` is always `Freeze + Copy`, regardless of `T`, so no qualif bits need to be set.

The only reason they were before was for promotion, which is why we split promotion into its own check.

Reply:
I think you're right. This means I can clean up a lot more!